pkg/catalog/catalog.go

```diff
@@ -1940,29 +1969,116 @@ func (c *Catalog) VerifyLinkAddress(ctx context.Context, repository, token strin
 }
 
 func (c *Catalog) DeleteExpiredLinkAddresses(ctx context.Context) {
-	it, err := c.Store.ListRepositories(ctx)
+	repos, err := c.listRepositoriesHelper(ctx)
 	if err != nil {
-		c.log.WithError(err).Warn("Failed to list repositories during delete expired addresses")
+		c.log.WithError(err).Warn("Failed list repositories during delete expired addresses")
 		return
 	}
+
+	for _, repo := range repos {
+		err := c.Store.DeleteExpiredLinkAddresses(ctx, repo)
+		if err != nil {
+			c.log.WithError(err).WithField("repository", repo.RepositoryID).Warn("Delete expired address tokens failed")
+		}
+	}
+}
+
+func (c *Catalog) listRepositoriesHelper(ctx context.Context) ([]*graveler.RepositoryRecord, error) {
+	it, err := c.Store.ListRepositories(ctx)
+	if err != nil {
+		return nil, err
+	}
 	defer it.Close()
 
 	var repos []*graveler.RepositoryRecord
 	for it.Next() {
-		record := it.Value()
-		repos = append(repos, record)
+		repos = append(repos, it.Value())
 	}
 	if err := it.Err(); err != nil {
-		c.log.WithError(err).Warn("Failed to iterate over repositories during delete expired addresses")
+		return nil, err
+	}
+	return repos, nil
+}
+
+func (c *Catalog) DeleteTrackedPhysicalAddresses(ctx context.Context) {
+	repos, err := c.listRepositoriesHelper(ctx)
+	if err != nil {
+		c.log.WithError(err).Warn("Failed list repositories during delete tracked physical addresses")
 		return
 	}
 
 	for _, repo := range repos {
-		err := c.Store.DeleteExpiredLinkAddresses(ctx, repo)
+		if err := c.deleteRepoTrackedPhysicalAddresses(ctx, repo); err != nil {
+			c.log.WithError(err).
+				WithField("repository", repo.RepositoryID).
+				Warn("Repository delete tracked physical addresses failed")
+		}
+	}
+}
+
+func (c *Catalog) deleteRepoTrackedPhysicalAddresses(ctx context.Context, repo *graveler.RepositoryRecord) error {
+	// collect and delete old keys - work in 'kvTrackDeleteBatchSize' and interval of 5 min between if batch is bigger
+	// based on the last modified field.
+	for {
+		hasMore, err := c.deleteRepoTrackedPhysicalAddressHelper(ctx, repo)
 		if err != nil {
-			c.log.WithError(err).WithField("repository", repo.RepositoryID).Warn("Delete expired address tokens failed")
+			return err
+		}
+		if !hasMore {
+			break
+		}
+		// sleep and process more
+		select {
+		case <-time.After(kvTrackDeleteBatchInterval):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
+func (c *Catalog) deleteRepoTrackedPhysicalAddressHelper(ctx context.Context, repo *graveler.RepositoryRecord) (bool, error) {
+	repoPartition := graveler.RepoPartition(repo)
+	msgType := (&Entry{}).ProtoReflect().Type()
+	it, err := kv.NewPrimaryIterator(ctx, c.kvStore.Store, msgType, repoPartition, []byte(kvTrackPrefix), kv.IteratorOptionsFrom([]byte{}))
```

Comment:
You're always starting from the top. 
Every 2 seconds you will be getting the same results which should not be deleted, it's a waste in kv calls

Reply:
Oh, is that the reason for the sorted keys?

Reply:
yes, I want to delete old 1k keys and repeat the action without overload the system with delete requests.